rs/replicated_state/src/canister_state/queues.rs

```diff
@@ -183,80 +135,77 @@ pub struct NewCanisterQueues {
-        memory_stats: &'a mut MemoryUsageStats,
-        queue_stats: &'a mut OutputQueuesStats,
+        queues: &'a mut BTreeMap<CanisterId, (CanisterQueue, CanisterQueue)>,
+        pool: &'a mut MessagePool,
     ) -> Self {
         let queues: VecDeque<_> = queues
             .iter_mut()
-            .filter(|(_, (_, queue))| queue.num_messages() > 0)
+            .filter(|(_, (_, queue))| queue.len() > 0)
             .map(|(canister, (_, queue))| (canister, queue))
             .collect();
         let size = Self::compute_size(&queues);
 
-        CanisterOutputQueuesIterator {
-            queues,
-            size,
-            memory_stats,
-            queue_stats,
-        }
+        CanisterOutputQueuesIterator { queues, pool, size }
     }
 
-    /// Returns a reference to the message that `pop` / `next` would return.
+    /// Returns the first message from the next queue.
     pub fn peek(&self) -> Option<&RequestOrResponse> {
-        if let Some((_, queue)) = self.queues.front() {
-            let msg = queue.peek().expect("Empty queue in iterator");
-            return Some(msg);
-        }
-        None
+        let item = self.queues.front()?.1.peek().unwrap();
+
+        let msg = self.pool.get(item.id());
+        assert!(msg.is_some(), "stale reference at the head of output queue");
+        msg
     }
 
-    /// Pops a message from the next queue. If this was not the last message in
-    /// that queue, the queue is moved to the back of the iteration order.
+    /// Pops the first message from the next queue.
+    ///
+    /// Advances the queue to the next non-stale message. If such a message exists,
+    /// the queue is moved to the back of the iteration order, else it is dropped.
     pub fn pop(&mut self) -> Option<RequestOrResponse> {
-        if let Some((receiver, queue)) = self.queues.pop_front() {
-            let msg = queue.pop().expect("Empty queue in iterator");
-
-            if queue.num_messages() > 0 {
-                self.queues.push_back((receiver, queue));
-            }
+        let (receiver, queue) = self.queues.pop_front()?;
+        debug_assert!(self.size >= queue.len());
+        self.size -= queue.len();
 
-            *self.memory_stats -= MemoryUsageStats::stats_delta(QueueOp::Pop, &msg);
-            *self.queue_stats -= OutputQueuesStats::stats_delta(&msg);
-            self.size -= 1;
-            debug_assert_eq!(Self::compute_size(&self.queues), self.size);
+        // Queue must be non-empty and message at the head of queue non-stale.
+        let msg = pop_and_advance(queue, self.pool).unwrap();
+        debug_assert_eq!(Ok(()), canister_queue_ok(queue, self.pool, receiver));
 
-            return Some(msg);
+        if queue.len() > 0 {
+            self.size += queue.len();
+            self.queues.push_back((receiver, queue));
         }
-        None
+        debug_assert_eq!(Self::compute_size(&self.queues), self.size);
+
```

Comment:
Should we also have debug asserts for the invariants here?

Reply:
We have 3 invariant check functions: `canister_queue_ok()` we call above; `stats_ok()` is kind of hard to call from here, but we're not touching slot or memory reservations while popping from output queues; and `schedules_ok()` is unnecessary, because we're not touching input queues at all.